In [2]:
import pandas as pd
import gurobipy as gp
from gurobipy import GRB
import scipy.sparse as sp
import numpy as np

In [3]:
brews_df = pd.read_pickle('data/brews_with_costs.pkl')
ingredients_df = pd.read_pickle('data/ingredients.pkl')
brews_df

first_ingredient       second_ingredient     third_ingredient  \
0      Abecean Longfin      Ancestor Moth Wing  Cyrodilic Spadetail   
1      Abecean Longfin       Ash Creep Cluster         Beehive Husk   
2      Abecean Longfin       Ash Creep Cluster         Chaurus Eggs   
3      Abecean Longfin       Ash Creep Cluster  Cyrodilic Spadetail   
4      Abecean Longfin       Ash Creep Cluster            Elves Ear   
...                ...                     ...                  ...   
25237    Tundra Cotton              Void Salts                   NA   
25238    Tundra Cotton          Wisp Wrappings                   NA   
25239     Vampire Dust               White Cap                   NA   
25240       Void Salts          Wisp Wrappings                   NA   
25241            Wheat  Yellow Mountain Flower                   NA   

                                      effects                            name  \
0       [Fortify Restoration, Damage Stamina]   Potion of Fortify Restoration   
1        [Fortify Sneak, Fortify Destruction]   Potion of Fortify Destruction   
2          [Weakness to Poison, Invisibility]          Potion of Invisibility   
3       [Fortify Restoration, Damage Stamina]   Potion of Fortify Restoration   
4            [Weakness to Frost, Resist Fire]           Potion of Resist Fire   
...                                       ...                             ...   
25237         [Resist Magic, Fortify Magicka]       Potion of Fortify Magicka   
25238                          [Resist Magic]          Potion of Resist Magic   
25239                       [Restore Magicka]       Potion of Restore Magicka   
25240                          [Resist Magic]          Potion of Resist Magic   
25241  [Fortify Health, Damage Stamina Regen]  Poison of Damage Stamina Regen   

       value                                       descriptions  
0        243  [Drain the target's Stamina by 22 points., Res...  
1        432  [Destruction spells are 46% stronger for 60 se...  
2        486  [Invisibility for 37 seconds., Target is 15% w...  
3        243  [Drain the target's Stamina by 22 points., Res...  
4        190  [Resist 28% of fire damage for 60 seconds., Ta...  
...      ...                                                ...  
25237    194  [Magicka is increased by 37 points for 60 seco...  
25238     80               [Resist 9% of magic for 60 seconds.]  
25239     52                    [Restore 58 points of Magicka.]  
25240     80               [Resist 9% of magic for 60 seconds.]  
25241    356  [Decrease the target's Stamina regeneration by...  

[25242 rows x 7 columns]

In [4]:
brews = list(brews_df.name)
ingredients = list(ingredients_df.name)

The model is the following:

#### Parameters

$B = \{1, \dots, 25242\}$: Set of brews

$I = \{1, \dots, 109\}$: Set of ingredients

$I_p \subset{I}$: Set of ingredients required to make one unit of brew $b$ (note that $|I_b| \leq 3 \,\,\forall b \in B$)

$z_i \in \mathbb{Z}^+$: Amount of ingredient $i$ available

$c_b \in \mathbb{R}^+$: gold value of brew $b$

#### Decision Variables

$y_b \in \mathbb{Z}^+$: Number of brews $b$ created

$x_{ib} \in \mathbb{Z}^+$: Amount of ingredient $i$ used to create brew $b$

#### Model

$$ \max \sum_{i \in b} c_b y_p$$
such that:
$$y_b \leq x_{ib} \qquad \forall i \in I_b,  \forall b \in B$$
$$\sum_{\{b \mid i \in I_b\}} x_{ib} \leq z_i \qquad \forall i \in I$$


An additional sophistication for the model is to allow the purchase of common ingredients, which would amount to adding a integer decision variable for each such ingredient, adding them to the RHSs of the second constraints, and subtracting the costs in the objective.

In [5]:
B = [i for i in range(len(brews_df))]
I = [i for i in range(len(ingredients_df))]
c = np.array(brews_df['value'])

In [6]:
A = [[] for b in B]
for b in B:
    brew_ingredients = brews_df.loc[b][['first_ingredient','second_ingredient','third_ingredient']]
    for ingredient in brew_ingredients:
        if ingredient != 'NA':
            i = ingredients.index(ingredient)
            A[b].append(i)
            
# randomly generate inventory
z = np.random.poisson(lam=1, size=len(I))

Using Gurobi:

In [7]:
l = gp.tuplelist()
for b in B:
    for i in A[b]:
        l.append((b,i))
        
m = gp.Model('alchemy')
x = m.addVars(l, vtype=GRB.CONTINUOUS, name='x')
y = m.addVars(B, vtype=GRB.INTEGER, name='y')
m.addConstrs((y[b] <= x[b,i] for (b,i) in l), 'c1')
m.addConstrs((x.sum('*', i) <= z[i] for i in I))

m.setObjective(gp.quicksum(c[b]*y[b] for b in B), GRB.MAXIMIZE)
m.setParam('OUTPUT_FLAG',False)
m.optimize()

Academic license - for non-commercial use only - expires 2022-07-08
Using license file C:\Users\Eric\gurobi.lic


In [10]:
optimal_brews = [b for b in B if y[b].x > 0.5]
optimal_df = brews_df.loc[optimal_brews]
optimal_df['count'] = [int(y[b].x) for b in B if y[b].x > 0.5]
optimal_df.sort_values(by=['value'], ascending=False)
optimal_df = optimal_df[['name', 'count', 'value', 'first_ingredient', 'second_ingredient', 'third_ingredient', 'descriptions']]
optimal_df = optimal_df.sort_values(by='value', ascending=False)
optimal_df

name  count  value         first_ingredient  \
7126        Potion of Fortify Stamina      1   6588                Boar Tusk   
23091                  Poison of Slow      1   5232            Large Antlers   
3010         Potion of Fortify Health      1   5145               Bear Claws   
16694          Potion of Invisibility      1    824    Felsaad Tern Feathers   
8296   Poison of Damage Magicka Regen      1    820      Burnt Spriggan Wood   
15183     Potion of Regenerate Health      1    804     Emperor Parasol Moss   
9257              Poison of Paralysis      1    701               Canis Root   
9557              Poison of Paralysis      1    693               Canis Root   
13276     Potion of Regenerate Health      1    691             Daedra Heart   
5249   Poison of Damage Magicka Regen      1    677              Blisterwort   
5504   Poison of Damage Magicka Regen      1    606      Blue Butterfly Wing   
7287   Poison of Damage Magicka Regen      1    589                Bone Meal   
10974  Poison of Damage Magicka Regen      1    580  Chaurus Hunter Antennae   
10527          Potion of Invisibility      1    572             Chaurus Eggs   
17702  Poison of Damage Stamina Regen      1    546            Frost Mirriam   
18076    Potion of Regenerate Magicka      1    536              Frost Salts   
16850    Potion of Regenerate Magicka      1    536               Fire Salts   
13298  Poison of Damage Stamina Regen      1    481             Daedra Heart   
17149    Potion of Regenerate Magicka      1    476               Fire Salts   
24354         Potion of Fortify Block      1    470                    Pearl   
4904          Potion of Fortify Block      1    447           Bleeding Crown   
3619     Potion of Regenerate Stamina      1    406                      Bee   
4568          Potion of Fortify Block      1    384           Bleeding Crown   
9565     Potion of Fortify One-handed      1    380               Canis Root   
21929  Poison of Damage Stamina Regen      1    373                Honeycomb   
1929     Potion of Fortify Pickpocket      1    373         Ash Hopper Jelly   
25009          Potion of Resist Frost      1    359               Scathecraw   
6044    Potion of Fortify Restoration      1    332            Blue Dartwing   
4221          Potion of Fortify Sneak      1    328             Beehive Husk   
21528  Poison of Damage Stamina Regen      1    294                 Histcarp   
3357                 Poison of Frenzy      1    259                      Bee   
3435          Potion of Resist Poison      1    242                      Bee   
3525      Poison of Weakness to Shock      1    216                      Bee   
2206    Potion of Fortify Light Armor      1    209         Ash Hopper Jelly   
14686       Potion of Fortify Magicka      1    170                Ectoplasm   
5179                 Poison of Frenzy      1    168              Blisterwort   
8909         Poison of Damage Magicka      1    118           Butterfly Wing   

             second_ingredient       third_ingredient  \
7126                    Garlic      Slaughterfish Egg   
23091             Poison Bloom             Trama Root   
3010                 Boar Tusk           Hanging Moss   
16694           Luna Moth Wing           Vampire Dust   
8296   Chaurus Hunter Antennae              Salt Pile   
15183                   Garlic              Troll Fat   
9257             Creep Cluster            Netch Jelly   
9557          Rock Warbler Egg       Swamp Fungal Pod   
13276          Juniper Berries           Namira's Rot   
5249                 Glow Dust            Human Heart   
5504   Chaurus Hunter Antennae                     NA   
7287   Chaurus Hunter Antennae             Hawk's Egg   
10974               Hawk's Egg               Lavender   
10527         Ice Wraith Teeth        Sabre Cat Tooth   
17702    Powdered Mammoth Tusk       Silverside Perch   
18076               Moon Sugar             Salmon Roe   
16850              Fros

In [11]:
print(f"Optimal Objective Value: {int(m.objVal)}")

Optimal Objective Value: 32625
